proof/refine/ARM/Ipc_R.thy

```diff
@@ -2304,20 +2296,478 @@ crunches doIPCTransfer, possibleSwitchTo
+                            tcb_at t and tcb_at a and scheduler_act_not a and
+                            (\<lambda>s. data \<noteq> None \<longrightarrow> reply_at (the data) s \<and>
+                                 ex_nonz_cap_to (the data) s \<and>
+                                 reply_tcb_reply_at (\<lambda>tptr. tptr = None) (the data) s \<and>
+                                 reply_sc_reply_at (\<lambda>tptr. tptr = None) (the data) s \<and>
+                                 the data \<notin> fst ` replies_with_sc s) and
+                            K (t \<noteq> idle_thread_ptr) and
+                            (\<lambda>s. cd \<longrightarrow> bound_sc_tcb_at (\<lambda>a. \<exists>y. a = Some y) t s)"
+                            in hoare_strengthen_post[rotated])
+                    apply (clarsimp cong: conj_cong)
+                    apply (frule valid_sched_action_weak_valid_sched_action, simp)
+                    apply (frule valid_objs_valid_tcbs, simp)
+                    apply (subgoal_tac "(cd \<longrightarrow> bound_sc_tcb_at (\<lambda>a. sc_at (the a) s) t s)")
+                     apply (clarsimp simp: obj_at_def is_tcb pred_tcb_at_def)
+                    apply (drule obj_at_ko_at, clarsimp)
+                    apply (frule (1) valid_objs_ko_at)
+                    apply (clarsimp simp: pred_tcb_at_def obj_at_def valid_obj_def valid_tcb_def)
+                   apply (wpsimp wp: reply_unlink_tcb_valid_sched_action
+                                     reply_unlink_tcb_valid_replies_BlockedOnReceive
+                                     reply_unlink_tcb_sym_refs_BlockedOnReceive
+                                     reply_unlink_tcb_reply_tcb_reply_at[where P=id, simplified]
+                                     reply_unlink_tcb_st_tcb_at[where P=id, simplified]
+                                     replies_with_sc_lift)
+                  apply (rule_tac Q="\<lambda>_. tcb_at' t and tcb_at' a and valid_release_queue and
+                                         valid_release_queue' and
+                                         valid_queues and valid_objs' and
+                                         valid_queues' and
+                                         (\<lambda>s. data \<noteq> None \<longrightarrow>
+                                              reply_at' (the data) s \<and>
+                                              replySCs_of s (the data) = None)"
+                                         in hoare_strengthen_post[rotated])
+                   apply (clarsimp cong: conj_cong)
+                   apply (frule valid_objs'_valid_tcbs')
+                   apply (drule obj_at_ko_at')+
+                   apply (blast dest: no_replySC_valid_replies'_sc_asrt)
+                  apply wpsimp
+                 apply (wpfix add: receive_reply_sel)
+                 apply (rule_tac Q="\<lambda>_. valid_objs and pspace_aligned and pspace_distinct and
```

Comment:
Can any of these large blocks be made simpler by using pattern matching against the statement of the lemma?